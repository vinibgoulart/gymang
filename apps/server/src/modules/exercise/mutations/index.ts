import ExerciseAdd from './ExerciseAddMutation';

const ExerciseMutations = {
  ExerciseAdd,
};

export default ExerciseMutations;
