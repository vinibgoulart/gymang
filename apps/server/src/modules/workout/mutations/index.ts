import WorkoutAdd from './WorkoutAddMutation';

const WorkoutMutations = {
  WorkoutAdd,
};

export default WorkoutMutations;
