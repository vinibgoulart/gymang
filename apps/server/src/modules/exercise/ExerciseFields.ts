import type { GraphQLContext } from '@gymang/core';
import { ExerciseFilterInputType, ExerciseLoader } from '@gymang/exercise';
import { edgeField, connectionArgs, withFilter } from '@gymang/graphql';
import { GraphQLNonNull } from 'graphql';
import type { ConnectionArguments } from 'graphql-relay';

import ExerciseType, { ExerciseConnection } from './ExerciseType';

export const exerciseTypeField = (
  key = 'exercise',
  bypassViewerCanSee = false,
) => ({
  [key]: {
    type: ExerciseType,
    resolve: async (obj, _, context: GraphQLContext) => {
      return ExerciseLoader.load(context, obj[key], bypassViewerCanSee);
    },
  },
});

export const exerciseEdgeField = () =>
  edgeField({
    connection: ExerciseConnection,
    load: ExerciseLoader.load,
    edgeName: 'exerciseEdge',
    name: 'Exercise',
  });

type MeExercisesConnectionArgs = {
  filters?: ExerciseFilterInputType;
} & ConnectionArguments;

export const exercisesConnectionField = <T = unknown>(
  customResolver = null,
) => ({
  exercises: {
    type: new GraphQLNonNull(ExerciseConnection.connectionType),
    args: {
      ...connectionArgs,
      filters: {
        type: ExerciseFilterInputType,
      },
    },
    resolve: (
      obj: T,
      args: MeExercisesConnectionArgs,
      context: GraphQLContext,
    ) => {
      if (customResolver) {
        return customResolver(obj, args, context);
      }

      const argsWithWorkoutSplit = withFilter(args, {
        workoutSplit: args.filters?.workoutSplit,
      });

      return ExerciseLoader.loadAll(context, argsWithWorkoutSplit);
    },
  },
});
