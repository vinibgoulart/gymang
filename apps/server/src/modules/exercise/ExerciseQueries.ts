import { exercisesConnectionField } from './ExerciseFields';

const ExerciseQueries = {
  ...exercisesConnectionField(),
};

export default ExerciseQueries;
