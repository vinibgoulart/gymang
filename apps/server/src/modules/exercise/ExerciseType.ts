import type { GraphQLContext } from '@gymang/core';
import type { IExercise } from '@gymang/exercise';
import { ExerciseLoader, ExerciseMuscleGroupEnum } from '@gymang/exercise';
import {
  connectionDefinitions,
  nodeInterface,
  registerTypeLoader,
} from '@gymang/graphql';
import {
  GraphQLBoolean,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLString,
} from 'graphql';
import { globalIdField } from 'graphql-relay';

import { userTypeField } from '../user/UserFields';
import { workoutSplitTypeField } from '../workoutSplit/WorkoutSplitFields';

const ExerciseType = new GraphQLObjectType<IExercise, GraphQLContext>({
  name: 'Exercise',
  description: 'Represents a Exercise',
  fields: () => ({
    id: globalIdField('Exercise'),
    name: {
      type: new GraphQLNonNull(GraphQLString),
      resolve: (exercise) => exercise.name,
    },
    series: {
      type: new GraphQLNonNull(GraphQLString),
      resolve: (exercise) => exercise.series,
    },
    repetitions: {
      type: new GraphQLNonNull(GraphQLString),
      resolve: (exercise) => exercise.repetitions,
    },
    breakTime: {
      type: GraphQLString,
      resolve: (exercise) => exercise.breakTime,
    },
    muscleGroup: {
      type: new GraphQLNonNull(ExerciseMuscleGroupEnum),
      resolve: (exercise) => exercise.muscleGroup,
    },
    weight: {
      type: GraphQLString,
      resolve: (exercise) => exercise.weight,
    },
    hasSessionInProgress: {
      type: GraphQLBoolean,
      resolve: (exercise) => {
        return !!exercise.sessions?.find(
          (session) => !session.finishedAt || session.finishedAt === null,
        );
      },
    },
    ...userTypeField(),
    ...workoutSplitTypeField(),
  }),
  interfaces: () => [nodeInterface],
});

export const ExerciseConnection = connectionDefinitions({
  name: 'Exercise',
  nodeType: ExerciseType,
});

registerTypeLoader(ExerciseType, ExerciseLoader.load);

export default ExerciseType;
