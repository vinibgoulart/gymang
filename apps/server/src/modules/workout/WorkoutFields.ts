import type { GraphQLContext } from '@gymang/core';
import { isLoggedIn } from '@gymang/core';
import {
  edgeField,
  connectionArgs,
  NullConnection,
  withFilter,
} from '@gymang/graphql';
import { WorkoutFilterInputType, WorkoutLoader } from '@gymang/workout';
import { GraphQLNonNull } from 'graphql';
import type { ConnectionArguments } from 'graphql-relay';

import WorkoutType, { WorkoutConnection } from './WorkoutType';

export const workoutTypeField = <T>(
  key = 'workout',
  bypassViewerCanSee = false,
) => ({
  [key]: {
    type: WorkoutType,
    resolve: async (obj: T, _, context: GraphQLContext) => {
      return WorkoutLoader.load(context, obj[key], bypassViewerCanSee);
    },
  },
});

export const workoutEdgeField = () =>
  edgeField({
    connection: WorkoutConnection,
    load: WorkoutLoader.load,
    edgeName: 'workoutEdge',
    name: 'Workout',
  });

type WorkoutConnectionArgs = {
  filters: WorkoutFilterInputType;
} & ConnectionArguments;

export const workoutConnectionField = <T>(customResolver = null) => ({
  workouts: {
    type: new GraphQLNonNull(WorkoutConnection.connectionType),
    args: {
      ...connectionArgs,
      filters: {
        type: WorkoutFilterInputType,
      },
    },
    resolve: (obj: T, args: WorkoutConnectionArgs, context: GraphQLContext) => {
      if (customResolver) {
        return customResolver(obj, args, context);
      }

      if (!isLoggedIn(context)) {
        return NullConnection;
      }

      const argsWithFilter = withFilter(args, {
        isPublic: true,
      });

      return WorkoutLoader.loadAll(context, argsWithFilter);
    },
  },
});

type MeWorkoutConnectionArgs = ConnectionArguments;

export const meWorkoutConnectionField = <T = unknown>(
  customResolver = null,
) => ({
  meWorkouts: {
    type: new GraphQLNonNull(WorkoutConnection.connectionType),
    args: {
      ...connectionArgs,
      // filters: {
      //   type: WorkoutFilterInputType,
      // },
    },
    resolve: (
      obj: T,
      args: MeWorkoutConnectionArgs,
      context: GraphQLContext,
    ) => {
      if (customResolver) {
        return customResolver(obj, args, context);
      }

      if (!isLoggedIn(context)) {
        return NullConnection;
      }

      const argsWithUser = withFilter(args, {
        user: context.user._id,
      });

      return WorkoutLoader.loadAll(context, argsWithUser);
    },
  },
});
