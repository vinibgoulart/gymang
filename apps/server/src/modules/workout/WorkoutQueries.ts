import { myWorkoutConnectionField } from './WorkoutFields';

const WorkoutQueries = {
  ...myWorkoutConnectionField(),
};

export default WorkoutQueries;
