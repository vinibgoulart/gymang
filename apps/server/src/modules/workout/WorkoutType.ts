import type { GraphQLContext } from '@gymang/core';
import {
  connectionDefinitions,
  nodeInterface,
  registerTypeLoader,
} from '@gymang/graphql';
import type { IWorkout } from '@gymang/workout';
import { WorkoutLoader } from '@gymang/workout';
import { GraphQLNonNull, GraphQLObjectType, GraphQLString } from 'graphql';
import { globalIdField } from 'graphql-relay';

import { userTypeField } from '../user/UserFields';

const WorkoutType = new GraphQLObjectType<IWorkout, GraphQLContext>({
  name: 'Workout',
  description: 'Represents a Workout',
  fields: () => ({
    id: globalIdField('Workout'),
    name: {
      type: new GraphQLNonNull(GraphQLString),
      resolve: (workout) => workout.name,
    },
    description: {
      type: GraphQLString,
      resolve: (workout) => workout.description,
    },
    ...userTypeField('createdBy'),
    ...userTypeField(),
  }),
  interfaces: () => [nodeInterface],
});

export const WorkoutConnection = connectionDefinitions({
  name: 'Workout',
  nodeType: WorkoutType,
});

registerTypeLoader(WorkoutType, WorkoutLoader.load);

export default WorkoutType;
