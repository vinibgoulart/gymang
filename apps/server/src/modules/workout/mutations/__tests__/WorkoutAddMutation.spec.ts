import { GRAPHQL_TYPE } from '@gymang/core';
import { Exercise, handleCreateExercise } from '@gymang/exercise';
import {
  clearDbAndRestartCounters,
  connectMongoose,
  disconnectMongoose,
  sanitizeTestObject,
} from '@gymang/testutils';
import { handleCreateUser } from '@gymang/user';
import { handleCreateWorkout } from '@gymang/workout';
import { WorkoutSplit, handleCreateWorkoutSplit } from '@gymang/workout-split';
import { graphql } from 'graphql';
import { toGlobalId } from 'graphql-relay';

import { getContext } from '../../../../getContext';
import { schema as schemaWeb } from '../../../../graphql/schema/schema';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

const query = `
  mutation WorkoutAddSpecMutation($input: WorkoutAddInput!) {
    WorkoutAdd(input: $input) {
      workout {
        id
        name
        createdBy {
          firstName
        }
        user {
          firstName
        }
      }
      success
      error
    }
  }
`;

it('should add a new workout', async () => {
  const user = await handleCreateUser();

  const input = {
    name: 'Chest Workout',
    isPublic: true,
  };

  const variables = {
    input,
  };

  const context = await getContext({
    graphql: GRAPHQL_TYPE.WEB,
    user,
  });

  const rootValue = {};

  const result = await graphql({
    schema: schemaWeb,
    source: query,
    rootValue,
    contextValue: context,
    variableValues: variables,
  });

  expect(result.errors).toBeUndefined();

  expect(result.data.WorkoutAdd.error).toBeNull();
  expect(result.data.WorkoutAdd.success).toEqual(
    'Workout created successfully',
  );

  const workoutCreated = result.data.WorkoutAdd.workout;

  expect(workoutCreated?.user.firstName).toEqual(user.firstName);
  expect(workoutCreated?.createdBy.firstName).toEqual(user.firstName);

  expect(sanitizeTestObject(result.data)).toMatchSnapshot();
});

it('should add a new workout created by other user and duplicate splits and exercises', async () => {
  const user = await handleCreateUser();
  const userWithWorkout = await handleCreateUser();

  const existentWorkout = await handleCreateWorkout({
    user: userWithWorkout,
    createdBy: userWithWorkout,
  });

  const workoutSplit = await handleCreateWorkoutSplit({
    workout: existentWorkout,
  });

  await handleCreateExercise({
    workoutSplit,
  });

  const input = {
    name: 'Chest Workout',
    isPublic: true,
    originalWorkout: toGlobalId('Workout', existentWorkout._id),
  };

  const variables = {
    input,
  };

  const context = await getContext({
    graphql: GRAPHQL_TYPE.WEB,
    user,
  });

  const rootValue = {};

  const result = await graphql({
    schema: schemaWeb,
    source: query,
    rootValue,
    contextValue: context,
    variableValues: variables,
  });

  expect(result.errors).toBeUndefined();

  expect(result.data.WorkoutAdd.error).toBeNull();
  expect(result.data.WorkoutAdd.success).toEqual(
    'Workout created successfully',
  );

  const workoutCreated = result.data.WorkoutAdd.workout;

  expect(workoutCreated?.user.firstName).toEqual(user.firstName);
  expect(workoutCreated?.createdBy.firstName).toEqual(
    userWithWorkout.firstName,
  );

  const allWorkoutSplits = await WorkoutSplit.find();

  expect(allWorkoutSplits.length).toEqual(2);

  const allExercises = await Exercise.find();

  expect(allExercises.length).toEqual(2);

  expect(sanitizeTestObject(result.data)).toMatchSnapshot();
});
