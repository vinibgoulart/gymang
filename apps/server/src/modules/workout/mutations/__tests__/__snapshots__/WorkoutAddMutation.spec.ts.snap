// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add a new workout 1`] = `
{
  "WorkoutAdd": {
    "error": "EMPTY",
    "success": "Workout created successfully",
    "workout": {
      "createdBy": {
        "firstName": "User 1 firstName",
      },
      "id": "FROZEN-ID",
      "name": "Chest Workout",
      "user": {
        "firstName": "User 1 firstName",
      },
    },
  },
}
`;

exports[`should add a new workout created by other user and duplicate splits and exercises 1`] = `
{
  "WorkoutAdd": {
    "error": "EMPTY",
    "success": "Workout created successfully",
    "workout": {
      "createdBy": {
        "firstName": "User 2 firstName",
      },
      "id": "FROZEN-ID",
      "name": "Chest Workout",
      "user": {
        "firstName": "User 1 firstName",
      },
    },
  },
}
`;
